v2/spanner-change-streams-to-sharded-file-sink/src/main/java/com/google/cloud/teleport/v2/templates/utils/ShardingLogicImplFetcher.java

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright (C) 2024 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.google.cloud.teleport.v2.templates.utils;
+
+import com.google.cloud.teleport.v2.spanner.migrations.schema.Schema;
+import com.google.cloud.teleport.v2.spanner.utils.IShardIdFetcher;
+import java.io.File;
+import java.io.IOException;
+import java.lang.reflect.Constructor;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.nio.channels.ReadableByteChannel;
+import java.nio.channels.WritableByteChannel;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.List;
+import org.apache.beam.sdk.io.FileSystems;
+import org.apache.beam.sdk.io.fs.ResourceId;
+import org.apache.beam.sdk.util.MimeTypes;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.base.Splitter;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.io.ByteStreams;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.io.Files;
+import org.joda.time.Duration;
+import org.joda.time.Instant;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Singleton class to get implementation of the shard identification logic. */
+public class ShardingLogicImplFetcher {
+  private static final Logger LOG = LoggerFactory.getLogger(ShardingLogicImplFetcher.class);
+  private static IShardIdFetcher shardIdFetcher = null;
+
+  public static synchronized IShardIdFetcher getShardingLogicImpl(
```

Comment:
This is called in the `@Setup` of the `AssignShardDoFn`, won't using `synchronized` mean that all parallel DoFn executions in different threads will block on this? 

Wondering if it will cause a performance impact 🤔

Please check this blog - 

https://medium.com/analytics-vidhya/apache-beam-a-minor-tip-for-cleaner-code-f293df3e3372

Reply:
No performance impact, ran a load test

Reply:
Can you evaluate the injection based approach covered in the blog?  `singleton` and `synchronised` in a massively parallel pipeline should be avoided as much as possible IMO.

Reply:
This cannot be injected for the simple reason that worker is in a different JVM. It was tried and it failed.
There locks are short held and will return the object immediately in most cases. Same thing was observed in load test.

Reply:
Interesting! 

My understanding of the [blog](https://medium.com/analytics-vidhya/apache-beam-a-minor-tip-for-cleaner-code-f293df3e3372) was - 

The first approach in the blog is similar to what is done in the PR - 

A `synchronised` block initializes the `RedistClient`, which is used the `ProcessElement` to fetch it.

The second approach in the blog then creates `RedisOptions` class which has a specific field called - `getRedissonClient()` - this field is annotated with a `@Default.InstanceFactory(RedisClientFactory.class)` which would use a factory to build the client.

As a result, the DoFn can cleanly get this object without the need to use `synchronisation`. Are you saying this won't work in the context of the change we are making here? Just like other `pipelineOptions` this is also just an object right (irrespective of how it was constructed).

I was thinking that analogously, there can be a factory class which builds the object of the class which is initialized using the JAR, and then that object can be injected? The `@Default.InstanceFactory(<class-name>.class)` would take care of building it 🤔

Please correct me if I am wrong!

Reply:
Could you tell how it will work for an object that is not in the same JVM?

Reply:
Discussed this offline, will try this out sometime on my own as a thought exercise.
Please feel free to go ahead with the current approach.